extension/wasm.cc

```diff
@@ -913,27 +1095,53 @@ PHP_MINIT_FUNCTION(wasm)
     return SUCCESS;
 }
 
-// Zend extension boilerplate.
+// Initialize the module information.
 PHP_MINFO_FUNCTION(wasm)
 {
     php_info_print_table_start();
     php_info_print_table_header(2, "wasm support", "enabled");
     php_info_print_table_end();
 }
 
+// Request initialization event.
+PHP_RINIT_FUNCTION(wasm)
+{
+#if defined(ZTS) && defined(COMPILE_DL_WASM)
+    ZEND_TSRMLS_CACHE_UPDATE();
+#endif
+
+    return SUCCESS;
+}
+
+// Request shutdown event.
+PHP_RSHUTDOWN_FUNCTION(wasm)
+{
+	return SUCCESS;
+}
+
+// Module shutdown event.
+PHP_MSHUTDOWN_FUNCTION(wasm)
+{
+    // Clean up persistent resources.
+    php_wasm_module_clean_up_persistent_resources();
```

Comment:
@nikic  When doing my test, I never saw the `wasm_module_destructor` function being executed, while this module shutdown was called (I tested with a basic `printf` to see it was executed). I found that the `persistent_list` hasmap was already empty when the module shutdown function was called. I tried to investigate and noticed this `zend_shutdown` function: If I read https://github.com/php/php-src/blob/787ecb6d567be5759d7b09d727afab5d1060adc8/Zend/zend.c#L1035-L1076 correctly, the `persistent_list` global variable is emptied by `zend_destroy_rsrc_list`, and then later the resource destructors are called by `zend_destroy_rsrc_list_dtors`. In the case of persistent resource, it means that the resource is registered in the `persistent_list`, and when the list is emptied by `zend_destroy_rsrc_list`, the associated resource destructors are not called.

Am I missing something? Am I doing something wrong here?

Apart from that, everything works great.

Reply:
@Hywan I don't think printf works during shutdown because the std file descriptors may already be closed. You need to open a handle to a file and print into it (or just use gdb...)

zend_destroy_rsrc_list should be calling the destructor, zend_destroy_rsrc_list_dtors destroys the destructor itself, it that makes sense ^^

Reply:
It makes sense! I was wondering about the standard file descriptors.

I did some test with a `fwrite` in `wasm_module_destructor`, and it does not seem to work. I tested with `php-cgi` for the SAPI. It is called however with `php-cli`.

Reply:
I don't see when `zend_destroy_rsrc_list` calls the resource destructors though 🤔.

Reply:
See https://github.com/wasmerio/php-ext-wasm/pull/34, I finally got it right.